src/3d/materials/qgsphongtexturedmaterialsettings.cpp

```diff
@@ -202,9 +246,29 @@ void QgsPhongTexturedMaterialSettings::addParametersToEffect( Qt3DRender::QEffec
 
   Qt3DRender::QParameter *ambientParameter = new Qt3DRender::QParameter( QStringLiteral( "ambientColor" ), ambientColor );
   Qt3DRender::QParameter *specularParameter = new Qt3DRender::QParameter( QStringLiteral( "specularColor" ), mSpecular );
-  Qt3DRender::QParameter *shininessParameter = new Qt3DRender::QParameter( QStringLiteral( "shininess" ), mShininess );
+  Qt3DRender::QParameter *shininessParameter = new Qt3DRender::QParameter( QStringLiteral( "shininess" ), static_cast<float>( mShininess ) );
 
   effect->addParameter( ambientParameter );
   effect->addParameter( specularParameter );
   effect->addParameter( shininessParameter );
 }
+
+void QgsPhongTexturedMaterialSettings::setDiffuseTexturePath( const QString &path )
+{
+  setDiffuseTexture( QVariant::fromValue( path ) );
+}
+
+void QgsPhongTexturedMaterialSettings::setDiffuseTexture( const QVariant &diffuse )
+{
+  Qt3DRender::QTexture2D *texture = diffuse.value<Qt3DRender::QTexture2D *>();
+  if ( texture )
+  {
+    mDiffuseTexturePath.clear();
+    mDiffuseTexture.reset( texture );
```

Comment:
I still don't get how this can be stable. QTexture2D is a QObject subclass, so we'll have both thread safety issues (if a QgsPhongTexturedMaterialSettings is copied across threads) + the risk of a parent object for the QTexture2D instance causing object deletion outside of the shared_ptr . 

@wonder-sk i'd love your insights here too, given that you're the resident expert in Qt3D internals :laughing:

Reply:
+1 to what Nyall said... passing QTexture2D around looks quite scary (and just like it was pointed out - QObject-based class in shared_ptr is not safe).

Looking at the QgsPhongTexturedMaterialSettings::toMaterial() implementation, we end up extracting QImage from it, so why not have `QgsPhongTexturedMaterialSettings::setDiffuseTexture( const QImage &diffuse )` instead? It would simplify things, not requiring two variants (either a path or a texture). Another argument against using QTexture2D is that PyQGIS users would not be able to use that (since we do not have PyQt3D linked).

Please double check also the functionality of terrain texture updates (e.g. when a map layer in project participating in the terrain's texture images gets enabled/disabled) - in such case, we do not recreate all the entities and materials again, instead we just update the image in `QgsTerrainTextureImage`. I believe this functionality is lost in the current PR, because the material now does not use the `QgsTerrainTextureImage` object (which gets new image), so this link seems to be lost. We may need some other mechanism to make sure that terrain's texture can be updated without a complete rebuild of the material...

Reply:
> +1 to what Nyall said... passing QTexture2D around looks quite scary (and just like it was pointed out - QObject-based class in shared_ptr is not safe).

This was never my intention to use a smart pointer, that was @nyalldawson's idea :laughing: !

> 
> Looking at the QgsPhongTexturedMaterialSettings::toMaterial() implementation, we end up extracting QImage from it, so why not have `QgsPhongTexturedMaterialSettings::setDiffuseTexture( const QImage &diffuse )` instead? It would simplify things, not requiring two variants (either a path or a texture). Another argument against using QTexture2D is that PyQGIS users would not be able to use that (since we do not have PyQt3D linked).
> 

Indeed. And that's the case now. 
Thanks for your comment. I came to the same conclusion. At first, I was trying to copy/paste Qt3D behavior but It creates a Material on the fly while QGIS stores settings and then creates a Material with `toMaterial()`. When I realized that the  `QImage from the stored `QTexture2D` needs to be copied to create a new `QTexture2D`, I thought that it would be much simpler to directly use a `QImage`. 

> Please double check also the functionality of terrain texture updates (e.g. when a map layer in project participating in the terrain's texture images gets enabled/disabled) - in such case, we do not recreate all the entities and materials again, instead we just update the image in `QgsTerrainTextureImage`. I believe this functionality is lost in the current PR, because the material now does not use the `QgsTerrainTextureImage` object (which gets new image), so this link seems to be lost. We may need some other mechanism to make sure that terrain's texture can be updated without a complete rebuild of the material...

I don't understand. On current master, on terrain texture updates, `QgsTerrainTileLoader::createTextureComponent` gets called and thus the material is recreated. I don't see how only the image is updated. Based on my understanding, this is is not the case on current master.
Anyway, the terrain texture updates works on this PR.

Reply:
Thanks for the updates!

> I don't understand. On current master, on terrain texture updates, QgsTerrainTileLoader::createTextureComponent gets called and thus the material is recreated. I don't see how only the image is updated. Based on my understanding, this is is not the case on current master.
Anyway, the terrain texture updates works on this PR.

Please try this use case:
1. open a 3D view
2. enable or disable a map layer (one that does not have a 3D renderer so that the terrain texture gets updated) - I expect this still works because QgsTextureMaterial + createTexture() are used
3. now in 3D view settings, enable terrain shading
4. enable or disable a map layer - here I am expecting that the terrain texture will not get updated anymore

The mechanism how the terrain texture is updated works like this (verified on qgis master):
- when the list of map layers for terrain texture gets modified, `QgsTerrainEntity::invalidateMapImages()` gets called
- this will create tile update jobs (`TerrainMapUpdateJob`) from `TerrainMapUpdateJobFactory` 
- when an updated texture is ready, `TerrainMapUpdateJob::onTileReady` is called and it uses `QgsTerrainTileEntity::textureImage()` to get `QgsTerrainTextureImage` and update the image

As far as I can tell, when terrain shading is enabled, `QgsTerrainTileLoader::createTextureComponent()` now does not use `createTexture()` which is the key bit here - it wraps texture's QImage into `QgsTerrainTextureImage` (which allows updates of the image without full re-creation of terrain's tile entity).

For terrain tile load, probably you may want to have an approach similar to QgsTextureMaterial::setTexture()... the `QgsPhongTexturedMaterialSettings` object would be created without any texture image, and in a second step you would pass the texture image wrapped into QgsTerrainTextureImage...

Reply:
> Thanks for the updates!
> 
> > I don't understand. On current master, on terrain texture updates, QgsTerrainTileLoader::createTextureComponent gets called and thus the material is recreated. I don't see how only the image is updated. Based on my understanding, this is is not the case on current master.
> > Anyway, the terrain texture updates works on this PR.
> 
> Please try this use case:
> 
>     1. open a 3D view
> 
>     2. enable or disable a map layer (one that does not have a 3D renderer so that the terrain texture gets updated) - I expect this still works because QgsTextureMaterial + createTexture() are used
> 
>     3. now in 3D view settings, enable terrain shading
> 
>     4. enable or disable a map layer - here I am expecting that the terrain texture will not get updated anymore
> 
> 
> The mechanism how the terrain texture is updated works like this (verified on qgis master):
> 
>     * when the list of map layers for terrain texture gets modified, `QgsTerrainEntity::invalidateMapImages()` gets called
> 
>     * this will create tile update jobs (`TerrainMapUpdateJob`) from `TerrainMapUpdateJobFactory`
> 
>     * when an updated texture is ready, `TerrainMapUpdateJob::onTileReady` is called and it uses `QgsTerrainTileEntity::textureImage()` to get `QgsTerrainTextureImage` and update the image
> 
> 
> As far as I can tell, when terrain shading is enabled, `QgsTerrainTileLoader::createTextureComponent()` now does not use `createTexture()` which is the key bit here - it wraps texture's QImage into `QgsTerrainTextureImage` (which allows updates of the image without full re-creation of terrain's tile entity).
> 
> For terrain tile load, probably you may want to have an approach similar to QgsTextureMaterial::setTexture()... the `QgsPhongTexturedMaterialSettings` object would be created without any texture image, and in a second step you would pass the texture image wrapped into QgsTerrainTextureImage...

Thanks for the explanation. I was unable to reproduce it because it only occurs when the terrain shading is enabled. 

I tried to find a solution with `QgsPhongTexturedMaterialSettings` but if we create it without a texture, it fallbacks to the non texture case which seems reasonable because we also need to handle the wrong texture path case. And we can't directly call a `setTexture` on the output of `toMaterial`. because it is merely a `QTextureMaterial`.  

So I ended up creating a new `QgsPhongTexturedMaterial` class which is mostly the code of   `QgsPhongTexturedMaterialSettings::toMaterial`.   
Now,  `QgsPhongTexturedMaterialSettings::toMaterial` calls `QgsPhongTexturedMaterial` with the good parameters and `QgsPhongTexturedMaterial` is directly used for the terrain with a `setTexture` call.    
This logic is similar to `QgsMetalRoughMaterial` / `QgsMetalRoughMaterialSettings`